### DISCUSSION TITLE
Berry Script - send telemetry via MQTT how to do?

### DISCUSSION THREAD

**Discussion by paulobriennz** [arendst/Tasmota#17667] [Q&A] [answered]
i have written a berry script that pulls data from a third party API for a webclient GET.  That bit working okay and code below.. what i am struggling with is the best way to 'publish' this so the Tasmota calls this script and takes the output and sends a regular MQTT message (like it does with STATE).

There isnt too much info on this, so trying to put together via different posts.. i see i can add a command 
https://tasmota.github.io/docs/Berry-Cookbook/

but trying to understand the best way to take the results in the ENERGY results below and upload MQTT message to: device/SENSOR

import json
import mqtt

def get_API(url)
  var cl = webclient()
  cl.begin(url)
  var r = cl.GET()
  if r != 200
    print('error getting data')
    return false
  end
  var s = cl.get_string()
  cl.close()
  return json.load(s) 
end

var topic = string.replace(string.replace(tasmota.cmd('FullTopic')['FullTopic'], '%topic%', tasmota.cmd('Topic')['Topic']), '%prefix%', tasmota.cmd('Prefix')['Prefix3']) + '/SENSOR'

var payload = {}
var energy = {}

var apiResult = get_API('http://1.1.1.1/api/getdata')

energy["Total"] = apiResult['total'] 
energy["Power"] = apiResult['power'] 
energy["Voltage"] = apiResult['voltage']
energy["Current"] = apiResult['current']

payload['Time'] = tasmota.time_str(tasmota.rtc()['local'])
payload['ENERGY'] = energy

#this is where i am manually publishing the MQTT message - but is there a better way - to create like a 'driver' that calls my function regularly and publishes MQTT like a normal hardware type telemetry event read from a hardware sensor??

mqtt.publish(topic, str(payload))

**Comment by barbudor** [comment, marked as answer]
The i2c sensor exemple [here](https://tasmota.github.io/docs/Berry-Cookbook/#creating-an-i2c-driver) give you some hints especially in step 7
Tasmota/berry will call the `append_json() ` method so your driver can add what is needed to the Sensor payload using `tasmota.response_append(msg)`

**Reply by sfromis** [reply]
To be clear, this strategy is that you do not control when you publish, but hook into the Tasmota + Berry framework for drivers providing data for the JSON sensor payloads and Tasmota front page. This is typically made for sensors polled about once a second, and with TelePeriod controlling how often data is actually published via MQTT. Usually this entails that you are not creating the timing loop yourself, but let Tasmota request data. Of course, you *can* still have your own timing loop, and have the driver callbacks provide data from that.

**Reply by paulobriennz** [reply]
awesome, cheers guys,.. that was good info.. solved my issue.. cheers
